### PR TITLE
(chore) - convert to urql and functional components

### PR DESCRIPTION
**Pull request by JoviDeCroock**
```  71.43 KB  build/static/js/2.ef450e01.chunk.js
  3.04 KB   build/static/js/main.4ada874f.chunk.js
  788 B     build/static/js/runtime~main.aab71a58.js
  296 B     build/static/css/main.39e772ee.chunk.css```

### CODE REVIEW

File: server/package.json [hunk 1]
```diff
@@ -8,7 +8,6 @@
   },
   "dependencies": {
     "bcryptjs": "^2.4.3",
-    "graphql": "^14.5.3",
```
Comment (JoviDeCroock) [inline server/package.json]: Apparently we have to respect the exact version used by yoga, else we get different realm errors

File: src/components/CreateLink.js [hunk 2]
```diff
@@ -15,57 +14,43 @@ const POST_MUTATION = gql`
...
-class CreateLink extends Component {
-  state = {
-    description: '',
-    url: '',
-  }
+const CreateLink = ({ history }) => {
+  const [description, setDescription] = React.useState("");
+  const [url, setUrl] = React.useState('');
+  const [state, executeMutation] = useMutation(POST_MUTATION);
+  const prevFetching = usePreviousValue(state.fetching);
```
Comment (JoviDeCroock) [inline src/components/CreateLink.js]: I replaced the onComplete hooks on `<Mutation>` with these

File: src/components/CreateLink.js [hunk 3]
```diff
@@ -15,57 +14,43 @@ const POST_MUTATION = gql`
...
 
-class CreateLink extends Component {
-  state = {
-    description: '',
-    url: '',
-  }
+const CreateLink = ({ history }) => {
+  const [description, setDescription] = React.useState("");
+  const [url, setUrl] = React.useState('');
+  const [state, executeMutation] = useMutation(POST_MUTATION);
```
Comment (JoviDeCroock) [inline src/components/CreateLink.js]: They pass their variables straight into this, I don't know if we want to do this too

File: src/components/Link.js [hunk 4]
```diff
@@ -23,45 +23,38 @@ const VOTE_MUTATION = gql`
...
 
-    return (
-      <div className="flex mt2 items-start">
-        <div className="flex items-center">
-          <span className="gray">{this.props.index + 1}.</span>
-          {authToken && (
-            <Mutation
-              mutation={VOTE_MUTATION}
-              variables={{ linkId: this.props.link.id }}
-              update={(store, { data: { vote } }) =>
```
Comment (JoviDeCroock) [inline src/components/Link.js]: Note that we don't seem to need this

File: src/components/LinkList.js [hunk 5]
```diff
@@ -76,126 +76,81 @@ const NEW_VOTES_SUBSCRIPTION = gql`
   }
 `
 
-class LinkList extends Component {
-  _updateCacheAfterVote = (store, createVote, linkId) => {
```
Comment (JoviDeCroock) [inline src/components/LinkList.js]: This is the props.updateCacheAfterVote from previous comment

File: src/components/LinkList.js [hunk 6]
```diff
@@ -76,126 +76,81 @@ const NEW_VOTES_SUBSCRIPTION = gql`
...
-            links: [newLink, ...prev.feed.links],
-            count: prev.feed.links.length + 1,
-            __typename: prev.feed.__typename,
-          },
-        })
-      },
-    })
-  }
-
-  _subscribeToNewVotes = subscribeToMore => {
```
Comment (JoviDeCroock) [inline src/components/LinkList.js]: TODO: test if this just works too

File: src/components/LinkList.js [hunk 7]
```diff
@@ -76,126 +76,81 @@ const NEW_VOTES_SUBSCRIPTION = gql`
...
-      query: FEED_QUERY,
-      variables: { first, skip, orderBy },
-    })
-
-    const votedLink = data.feed.links.find(link => link.id === linkId)
-    votedLink.votes = createVote.link.votes
-    store.writeQuery({ query: FEED_QUERY, data })
-  }
-
-  _subscribeToNewLinks = subscribeToMore => {
```
Comment (JoviDeCroock) [inline src/components/LinkList.js]: TODO: test if this just works too

File: src/components/Search.js [hunk 8]
```diff
@@ -26,39 +26,25 @@ const FEED_SEARCH_QUERY = gql`
...
 
-class Search extends Component {
-  state = {
-    links: [],
-    filter: '',
-  }
+const Search = () => {
+  const [filter, setFilter] = React.useState('');
+
+  const [result] = useQuery({ query: FEED_SEARCH_QUERY, variables: { filter} });
```
Comment (JoviDeCroock) [inline src/components/Search.js]: They only query on button click but personally I like this more

File: src/index.js [hunk 9]
```diff
@@ -3,60 +3,90 @@ import ReactDOM from 'react-dom'
...
-const authLink = setContext((_, { headers }) => {
-  const token = localStorage.getItem(AUTH_TOKEN)
-  return {
-    headers: {
-      ...headers,
-      authorization: token ? `Bearer ${token}` : '',
-    },
+const cache = cacheExchange({
+  keys: {
+    Feed: data => data.id || data._id || null,
```
Comment (JoviDeCroock) [inline src/index.js]: I had to add keys for moments where we send a Mutation without `id` in it

File: src/index.js [hunk 10]
```diff
@@ -3,60 +3,90 @@ import ReactDOM from 'react-dom'
...
+    Link: data => data.id || data._id || null
+  },
+  updates: {
+    Mutation: {
+      post: (result, args, cache) => {
+        // This doesn't seem needed...
```
Comment (JoviDeCroock) [inline src/index.js]: The two `update` that we don't need

File: server/src/resolvers/Mutation.js [hunk 11]
```diff
@@ -2,10 +2,16 @@ const bcrypt = require('bcryptjs')
 const jwt = require('jsonwebtoken')
 const { APP_SECRET, getUserId } = require('../utils')
 
-function post(parent, args, context) {
+function post(parent, { url, description }, context) {
+  const userId = getUserId(context)
```
Comment (kitten) [inline server/src/resolvers/Mutation.js]: I think we'll have to be a little clever and make a separate branch where this change hasn't been made for the tutorial intro 🤔

Comment (JoviDeCroock) [reply server/src/resolvers/Mutation.js]: We might just opt to not throw in our final solution?

Comment (kitten) [reply server/src/resolvers/Mutation.js]: I think it's safer if we update the starter branch with the basic stuff from after the getting started step. Happy to do that once we merge this PR :)

Comment (JoviDeCroock) [reply server/src/resolvers/Mutation.js]: well for starter we only need to remove this line. That will be used for codeload right?

File: src/index.js [hunk 12]
```diff
@@ -21,33 +21,38 @@ import { FEED_QUERY } from './components/LinkList';
 import { getToken } from './token';
 
 const cache = cacheExchange({
+  keys: {
```
Comment (kitten) [inline src/index.js]: Let’s not add that and have the warnings for now since they’re safe anyway :)

**Pull request merged**